![mycotools](https://gitlab.com/xonq/mycotools/-/raw/master/misc/ablogo.png) ![mycotools database](https://gitlab.com/xonq/mycotools/-/raw/master/misc/mtdb.png)

# Mycotools Database (MTDB)

MTDBs are locally assimilated, uniformly curated databases of JGI MycoCosm (preferred) and NCBI
fungal or prokaryotic genomic data. MTDBs are represented in tab delimitted database `.mtdb` reference files,
which serve as the scaleable input to Mycotools scripts. 

Herein are the objectives, standards, and expectations of MTDB and associated files.

<br /><br />

## OBJECTIVE

Enable broadscale comparative genomics via a systematically curated, automatically
assembled/updated, scaleable genomics database. MTDB primarily seeks to resolve
several outstanding problems in comparative genomics: 

1. Uniformly curate genomic data within and across multiple databases, i.e. the notorious
inconsistency of the gene coordinates, `gff` file
2. Promote ease-of-use for scalable and large scale analyses, e.g. transitioning between datasets
in a phylogenetic analysis
3. Keep-up with the accelerating deposition of public genomic data via automatic updates
4. Implement a modular comparative genomic analyses toolkit to enable automated pipelining and 
make routine comparative genomic analyses accessible

<br /><br /><br />

## MTDB standards
### `.mtdb` file format standard
Example row:
```
#ome    genus   species strain  taxonomy        version source  biosample   assembly_acc    acquisition_date        published       fna     faa     gff3
aaoarx1 Aaosphaeria     arxii   CBS17579        {"clade": "dothideomyceta", "kingdom": "Fungi", "phylum": "Ascomycota", "subphylum": "Pezizomycotina", "class": "Dothideomycetes", "subclass": "Pleosporomycetidae", "order": "Pleosporales", "family": "", "subfamily": "", "genus": "Aaosphaeria", "species": "Aaosphaeria arxii"}    v1.0    jgi             Aaoar1  20210414 Haridas S et al.,2020
```

Tab-delimited file, with one row per genome and ordered columns:
- `ome`: MTDB accession "ome" - first three letters of genus, first three letters of species
(or "sp."), unique database number, and optional MTDB version tag '.\d+', e.g.
`psicub1`/`cryneo24.1` `[a-zA-Z0-9\.]`
- `genus`: Genus name from NCBI/Mycocosm/inputted master table; `[a-zA-Z]`
- `species`: Species name from NCBI/Mycocosm/inputted master table; `[a-zA-Z]`
- `strain`: Strain name with only letters and numbers from NCBI/Mycocosm master table `[a-zA-z0-9]`
- `taxonomy`: NCBI taxonomy `JSON` object derived from genus
- `version`: Mycocosm version/NCBI modification date
- `source`: Genome source, e.g. 'ncbi'/'jgi'/'lab'; `[a-z0-9]`
- `biosample`: NCBI biosample accession
- `assembly_acc`: NCBI GenBank/RefSeq assembly accession or Mycocosm genome_code
- `published`: Publication metadata or binary publication response
- `acquisition_date`: Data of input into master database `YYYYmmdd`
- `fna`: assembly `.fna`, required when not `$MYCOFNA/fna/<ome>.fna`
- `faa`: proteome `.faa`, required when not `$MYCOFAA/faa/<ome>.faa`
- `gff3`: gene coordinate `.gff3`, required when not `$MYCOGFF3/gff3/<ome>.gff3`

If headers are included, the line must beging with '#'

MTDB requires an assembly and gene coordinates `gff3` for ALL GENOMES.
Proteomes will be generated by referencing the assembly and `gff3`.

<br /><br />

### accession formatting
All MTDB aliases will be formatted as `<ome>_<acc>` where `ome` is `ome` in the
MTDB and acc is the retrieved accession for both assemblies and proteomes.
Thus, MTDB aliases are directly connected to the MTDB by slicing to the
underscore.
For JGI, accessions will pull from the `protein_id` field in the gene coordinates file.
For NCBI, accessions will pull from the `product_id` field in the gene coordinates file.
For entries without a detected protein ID, an alias will be assigned with the
prefix, 'mtdb'. Pseudogenes, tRNAs, and rRNAs aliases will format as
`<ome>_<type><type_count>`

<br /><br />

### `gff3` gene coordinates file standard
MTDB attempts to curate, assimilate, and modernize ALL MycoCosm and NCBI
annotations, including legacy data. All proteins, associated transcripts, 
exons, gene, and CDSs will include `;Alias=<ome>_<acc>` in the attributes
column. 
All attribute fields will contain `ID=[^;]+`, `Alias=[^;]+`; 
Non-gene entries will have a parent field `Parent=[^;]+` that relates the entry
to its parent RNA and each RNA to their parent gene.

On the occassion GFF entries are not given an attribute, *assume that these are
ignored by Mycotools*; while curation is fairly robust for JGI and NCBI GFFs,
non-permitted gff sequence fields that pull through curation may not receive an alias
and are likely ignorable. CDSs without an alias will not be translated into the
proteome `faa` fasta.

#### - permitted `gff3` sequence type fields: 
- gene, pseudogene: contains the terminal ID of descendant entries and alias (Alias=.*;) that contains
  all MTDB aliases derived from the gene, separated by `|`
- mRNA, tRNA, rRNA, RNA: `RNA` is synonymous with `transcript` and represents
  transcribed, but not translated, sequences without tRNA/rRNA functionality
- exon: exon ID will contain a unique number `exon\d+`; parent will be an RNA
  ID; introns will be curated to exons
- CDS: CDS ID will contain a unique number `cds\d+`; parent will be an RNA ID; typically contains a
  `protein_id/product_id` field

#### - alternate splicing
Alternately spliced genes are accounted for in curation. Genes with alternately
spliced descendants will have multiple aliases, separated by '|'. mRNAs and
their children will all have unique aliases.

#### - proteome `faa`
Proteomes will be generated on the fly when updating the database by
referencing an MTDB-curated `gff3` and assembly. Proteins are generated from
CDS coordinates that can be tied to an mRNA with a gene parent. 

<br /><br />

### data assimilation
#### - JGI
`updateDB.py` will prioritize Mycocosm (JGI) genomes over NCBI by referencing
the submitter field in NCBI assembly metadata. Each unique Portal is retrieved from the
MycoCosm master table.

#### - NCBI
NCBI genomes will be retrieved from the master eukaryotes.txt/prokaryotes.txt 
and each unique assembly accession is retrieved. BioSample accessions
that overlap with entries from Mycocosm are removed before assimilation.
Version checking operates on the `Modify Date` field. All NCBI entries are
assumed to be "published" for non-restricted use.

#### - local
Locally annotated genomes can be added to the database by filling out and
submitting a `.predb` file using `predb2db.py`. `predb2db.py` will curate the
inputted data and output into the current directory. Once complete,
`updateDB.py -a <PREDB_RESULT>` will add the `.mtdb` from `predb2db.py` to the
master database.

<br /><br /><br />

## Database management

The master MTDB should be generated from one user, and privileges should be
distributed using `chmod`. Note, the master user/group are the only ones with 
privileges to update and merge manually curated `predb` files into the master database.
